x-pack/plugins/security_solution/server/lib/detection_engine/migrations/helpers.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { BadRequestError } from '../errors/bad_request_error';
+import { MigrationDetails, MigrationStatus } from './types';
+
+const decodeBase64 = (base64: string) => Buffer.from(base64, 'base64').toString('utf8');
+const encodeBase64 = (utf8: string) => Buffer.from(utf8, 'utf8').toString('base64');
+
+export const encodeMigrationToken = (details: MigrationDetails): string =>
+  encodeBase64(JSON.stringify(details));
+
+export const decodeMigrationToken = (token: string): MigrationDetails => {
+  try {
+    const details = JSON.parse(decodeBase64(token)) as MigrationDetails;
+
+    if (details.destinationIndex == null || details.sourceIndex == null || details.taskId == null) {
+      throw new TypeError();
+    }
+
+    return details;
+  } catch (_) {
+    throw new BadRequestError(`An error occurred while decoding the migration token: [${token}]`);
+  }
+};
+
+export const isOutdated = ({ current, target }: { current: number; target: number }): boolean =>
+  current < target;
+
+const mappingsAreOutdated = ({
+  status,
+  version,
+}: {
+  status: MigrationStatus;
+  version: number;
+}): boolean => isOutdated({ current: status.version, target: version });
+
+const signalsAreOutdated = ({
+  status,
+  version,
+}: {
+  status: MigrationStatus;
+  version: number;
+}): boolean =>
+  status.signal_versions.some((signalVersion) => {
+    return (
+      signalVersion.doc_count > 0 && isOutdated({ current: signalVersion.key, target: version })
+    );
+  });
```

Comment:
I believe that this could happen as part of a multi-kibana setup, where one instance has been upgraded but another is still generating signals on an old version (and both share the same signals index).

In those situations, I believe the migration recommendation will be to perform the migration on one instance after disabling jobs in all other instances, and then iteratively upgrading/re-enabling jobs on each subsequent instance.

Reply:
Regardless of _how_ outdated signals make it into one of these indexes, they have the potential to break certain features, and the check is straightforward, so I decided to notify the user of that situation.